repository/src/main/java/org/apache/atlas/accesscontrol/persona/callable/CleanRoleWorker.java

```
@@ -0,0 +1,118 @@
+    private static final Logger LOG = LoggerFactory.getLogger(CleanRoleWorker.class);
+
+    private RangerPolicy rangerPolicy;
+    private AtlasEntity persona;
+    private List<String> removePolicyGuids;
+    private AtlasRangerService atlasRangerService;
+
+    public CleanRoleWorker(AtlasEntity persona, RangerPolicy rangerPolicy, List<String> removePolicyGuids, AtlasRangerService atlasRangerService) {
+        this.rangerPolicy = rangerPolicy;
+        this.persona = persona;
+        this.removePolicyGuids = removePolicyGuids;
+        this.atlasRangerService = atlasRangerService;
+    }
+
+    @Override
+    public RangerPolicy call() {
+        RangerPolicy ret = null;
+        LOG.info("Starting CleanRoleWorker");
+
+        try {
+            boolean deletePolicy = false;
+            String role = AtlasPersonaUtil.getRoleName(persona);
+
+            if (rangerPolicy.getPolicyType() == RangerPolicy.POLICY_TYPE_ACCESS) {
+                deletePolicy = cleanRoleFromAccessPolicy(role, rangerPolicy);
+            } else {
+                deletePolicy = cleanRoleFromMaskingPolicy(role, rangerPolicy);
+            }
+
+            if (deletePolicy) {
+                atlasRangerService.deleteRangerPolicy(rangerPolicy);
+            } else {
+                rangerPolicy.getPolicyLabels().remove(AtlasPersonaUtil.getPersonaLabel(persona.getGuid()));
+                rangerPolicy.getPolicyLabels().removeAll(removePolicyGuids);
+
+                long policyLabelCount = rangerPolicy.getPolicyLabels().stream().filter(x -> x.startsWith(LABEL_PREFIX_PERSONA)).count();
+                if (policyLabelCount == 0) {
+                    rangerPolicy.getPolicyLabels().remove(LABEL_TYPE_PERSONA);
+                }
+
+                atlasRangerService.updateRangerPolicy(rangerPolicy);
+            }
+        } catch (AtlasBaseException e) {
+            LOG.error("Failed to clean Ranger role from Ranger policies: {}", e.getMessage());
```

api should throw an error in this case